backend/controller/ingress/ingress.go

```diff
@@ -0,0 +1,233 @@
+package ingress
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/TBD54566975/ftl/backend/common/slices"
+	"github.com/TBD54566975/ftl/backend/controller/dal"
+	"github.com/TBD54566975/ftl/backend/schema"
+)
+
+type path []string
+
+func (p path) String() string {
+	return strings.TrimLeft(strings.Join(p, ""), ".")
+}
+
+func GetIngressRoute(routes []dal.IngressRoute, method string, path string) (*dal.IngressRoute, error) {
+	var matchedRoutes = slices.Filter(routes, func(route dal.IngressRoute) bool {
+		return matchAndExtractAllSegments(route.Path, path, func(segment, value string) {})
+	})
+
+	if len(matchedRoutes) == 0 {
+		return nil, dal.ErrNotFound
+	}
+
+	// TODO: add load balancing at some point
+	route := matchedRoutes[rand.Intn(len(matchedRoutes))] //nolint:gosec
+	return &route, nil
+}
+
+func matchAndExtractAllSegments(pattern, urlPath string, onMatch func(segment, value string)) bool {
+	patternSegments := strings.Split(strings.Trim(pattern, "/"), "/")
+	urlSegments := strings.Split(strings.Trim(urlPath, "/"), "/")
+
+	if len(patternSegments) != len(urlSegments) {
+		return false
+	}
+
+	for i, segment := range patternSegments {
+		if segment == "" && urlSegments[i] == "" {
+			continue // Skip empty segments
+		}
+
+		if strings.HasPrefix(segment, "{") && strings.HasSuffix(segment, "}") {
+			key := strings.Trim(segment, "{}") // Dynamic segment
+			onMatch(key, urlSegments[i])
+		} else if segment != urlSegments[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func ValidateAndExtractBody(route *dal.IngressRoute, r *http.Request, sch *schema.Schema) ([]byte, error) {
+	requestMap, err := buildRequestMap(route, r)
+	if err != nil {
+		return nil, err
+	}
+
+	verb := sch.ResolveVerbRef(&schema.VerbRef{Name: route.Verb, Module: route.Module})
+	if verb == nil {
+		return nil, fmt.Errorf("unknown verb %s", route.Verb)
+	}
+
+	dataRef := verb.Request
+	if dataRef.Module == "" {
+		dataRef.Module = route.Module
+	}
+
+	err = validateRequestMap(dataRef, []string{dataRef.String()}, requestMap, sch)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := json.Marshal(requestMap)
+	if err != nil {
+		return nil, err
+	}
+
+	return body, nil
+}
+
+func buildRequestMap(route *dal.IngressRoute, r *http.Request) (map[string]any, error) {
+	requestMap := map[string]any{}
+	matchAndExtractAllSegments(route.Path, r.URL.Path, func(segment, value string) {
+		requestMap[segment] = value
+	})
+
+	switch r.Method {
+	case http.MethodPost, http.MethodPut:
+		var bodyMap map[string]any
+		err := json.NewDecoder(r.Body).Decode(&bodyMap)
+		if err != nil {
+			return nil, err
```

Comment:
I think add an `fmt.Errorf("HTTP request body is not valid JSON: %w", err)` so the returned error is a bit more useful.